common/persistence/sql/sqlExecutionStore.go

```diff
@@ -255,116 +256,107 @@ func (m *sqlExecutionStore) GetWorkflowExecution(
+	wfID := request.Execution.WorkflowID
 
-	{
+	var executionsError error
+	var executions []sqlplugin.ExecutionsRow
+	var activityInfos map[int64]*p.InternalActivityInfo
+	var timerInfos map[string]*p.TimerInfo
+	var childExecutionInfos map[int64]*p.InternalChildExecutionInfo
+	var requestCancelInfos map[int64]*p.RequestCancelInfo
+	var signalInfos map[int64]*p.SignalInfo
+	var bufferedEvents []*p.DataBlob
+	var signalsRequested map[string]struct{}
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	g.Go(func() error {
+		executions, executionsError = m.getExecutions(ctx, request, domainID, wfID, runID)
+		return executionsError
+	})
+
+	g.Go(func() error {
 		var err error
-		state.ActivityInfos, err = getActivityInfoMap(
-			ctx,
-			m.db,
-			m.shardID,
-			domainID,
-			wfID,
-			runID,
-			m.parser)
-		if err != nil {
-			return nil, err
-		}
-	}
+		activityInfos, err = getActivityInfoMap(
+			ctx, m.db, m.shardID, domainID, wfID, runID, m.parser)
+		return err
+	})
 
-	{
+	g.Go(func() error {
 		var err error
-		state.TimerInfos, err = getTimerInfoMap(
-			ctx,
-			m.db,
-			m.shardID,
-			domainID,
-			wfID,
-			runID,
-			m.parser)
-		if err != nil {
-			return nil, err
-		}
-	}
+		timerInfos, err = getTimerInfoMap(
+			ctx, m.db, m.shardID, domainID, wfID, runID, m.parser)
+		return err
+	})
 
-	{
+	g.Go(func() error {
 		var err error
-		state.ChildExecutionInfos, err = getChildExecutionInfoMap(
-			ctx,
-			m.db,
-			m.shardID,
-			domainID,
-			wfID,
-			runID,
-			m.parser)
-		if err != nil {
-			return nil, err
-		}
-	}
+		childExecutionInfos, err = getChildExecutionInfoMap(
+			ctx, m.db, m.shardID, domainID, wfID, runID, m.parser)
+		return err
+	})
 
-	{
+	g.Go(func() error {
 		var err error
-		state.RequestCancelInfos, err = getRequestCancelInfoMap(
-			ctx,
-			m.db,
-			m.shardID,
-			domainID,
-			wfID,
-			runID,
-			m.parser)
-		if err != nil {
-			return nil, err
-		}
-	}
+		requestCancelInfos, err = getRequestCancelInfoMap(
+			ctx, m.db, m.shardID, domainID, wfID, runID, m.parser)
+		return err
+	})
 
-	{
+	g.Go(func() error {
 		var err error
-		state.SignalInfos, err = getSignalInfoMap(
-			ctx,
-			m.db,
-			m.shardID,
-			domainID,
-			wfID,
-			runID,
-			m.parser)
-		if err != nil {
-			return nil, err
-		}
-	}
+		signalInfos, err = getSignalInfoMap(
+			ctx, m.db, m.shardID, domainID, wfID, runID, m.parser)
+		return err
+	})
 
-	{
+	g.Go(func() error {
 		var err error
-		state.BufferedEvents, err = getBufferedEvents(
-			ctx,
-			m.db,
-			m.shardID,
-			domainID,
-			wfID,
-			runID)
-		if err != nil {
-			return nil, err
+		bufferedEvents, err = getBufferedEvents(
+			ctx, m.db, m.shardID, domainID, wfID, runID)
+		return err
+	})
+
+	g.Go(func() error {
+		var err error
+		signalsRequested, err = getSignalsRequested(
+			ctx, m.db, m.shardID, domainID, wfID, runID)
+		return err
+	})
+
+	err := g.Wait()
+	if executionsError != nil {
+		return nil, executionsError
+	}
```

Comment:
I'm guessing this err is more likely to be relevant, hence it's preferred?

If so: I don't think that'll actually work.  Once anything in the errgroup fails, the context will be canceled, so ^ the executions error is very likely to be either A) the same as the errgroup err, because it failed first, or B) a context-cancelled err, because it failed after some other request.  Context-cancelled is almost always going to be worse than whatever the errgroup returns, since there's no information about the cause of the cancellation.

If we don't mind _always waiting_ for executions, it could just use the non-errgroup context.  `g.Wait()` will still wait for it to complete (it waits for all funcs to return), but that way it should have a relevant error, at the cost of potentially-increased run-time.

Reply:
Yeah this is a bit racy but dangerous IMO. We are paying for missing some of these errors (but get new ones) to get better performance. Given we will still get "some" error and this function is supposed to succeed 99+% of the time. I'm not really worried about it. I believe this is a good trade-off. Let me know if you'd like discuss about this.

Reply:
I don't mean to imply it's racy (it isn't, in the data-safety sense at least).  I mean that this should _almost always_ be the same as **or worse** than not tracking `executionsError` separately at all.

When something else fails, `executionsError` will _almost always_ be a context-canceled error because much of its time is spent doing RPC calls... which would be returned here, instead of the error from the thing-that-failed-first.  So this will _mostly_ suppress other useful errors, and return useless ones.

...with a very small window for races where two goroutines both fail after their associated requests are completed, but before they return (on the order of microseconds).  We might be able to capture those with a "... and executions err is not a canceled-context error", but I'd only expect it to hit that branch _much_ less than 1% of the time.